parts/xmem_nozzle.go

```diff
@@ -1442,6 +1429,25 @@ func (xmem *XmemNozzle) preprocessMCRequest(req *base.WrappedMCRequest, lookup *
 	return nil
 }
 
+func (xmem *XmemNozzle) setCasLocking(req *base.WrappedMCRequest, lookup *base.SubdocLookupResponse, opt SetMetaXattrOptions) {
+	if lookup == nil || lookup.Resp.Status == mc.KEY_ENOENT {
+		req.Req.Cas = 0
+	} else {
+		// Cas locking, meaning the request will only succeed if the target Cas matches this value
+		req.Req.Cas = lookup.Resp.Cas
+	}
+	if opt.noTargetCR {
+		if len(req.Req.Extras) < 28 {
+			extras := make([]byte, 28)
+			copy(extras, req.Req.Extras)
+			req.Req.Extras = extras
+		}
+		options := binary.BigEndian.Uint32(req.Req.Extras[24:28])
+		options |= base.SKIP_CONFLICT_RESOLUTION_FLAG
+		binary.BigEndian.PutUint32(req.Req.Extras[24:28], options)
```

Comment:
Can these be moved under a method
```
func (req *WrappedMCRequest) SetSetWithMetaFlags(flags []uint32) {
...
}
```
There are many other flags like `REGENERATE_CAS` or `FORCE_ACCEPT_WITH_META_OPS`, etc can just use this one instead.

Do we still need to do duplicate work in router then?
```		
if router.sourceCRMode == base.CRMode_LWW {
			// if source bucket is of lww type, add FORCE_ACCEPT_WITH_META_OPS options for memcached
			options |= base.FORCE_ACCEPT_WITH_META_OPS
		} else if router.sourceCRMode == base.CRMode_Custom {
			// Custom conflict resolution behaves the same as LWW in KV.
			// For mobile, we also skip target conflict resolution. This will be done in xmem based on
			// whether the mutation is being sent to target with mobile enabled. Setting it here can have
			// race condition since the setting cab change in the middle of the replication.
			options |= base.SKIP_CONFLICT_RESOLUTION_FLAG
			options |= base.FORCE_ACCEPT_WITH_META_OPS
		}

```

Should we consolidate the logic somewhere? Is it possible to do it either in the router or XMEM?

Reply:
It is more efficient to set it in router, since router can just allocate the right Extras size, and set the bytes. Most time Xmem can leave it unchanged. The only time Xmem will need to expand Extras from 24 bytes to 28 bytes is: router allocates 24 bytes (revId CR, not expiration), and Xmem needs to set SKIP_CONFLICT_RESOLUTION_FLAG when mobile==active. Moving all to XMEM means XMEM will always allocate a new Extras if it is LWW. It will impact LWW performance.